lib/waiter.js

```diff
@@ -0,0 +1,60 @@
+class Waiter {
+  constructor (promise, cb, expectReject) {
+    this.cb = cb
+    this.ready = false
+    this.value = null
+    this.resolved = false
+    this.rejected = false
+    this.done = false
+    this.finishing = false
+    this.expectReject = !!expectReject
+    this.promise = new Promise(res => this.resolve = res)
+    promise.then(value => {
+      if (this.done) {
+        return
+      }
```

Comment:
This check is newly added.  If `.abort()` happens when `promise` is resolved but the resolve is not yet received then we would eventually overwrite statuses / values with the `resolved` info.

Not sure if this can be when lib/test.js uses `new Waiter` but it can happen in isolated testing.